engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -432,4 +496,66 @@ private DirectBuffer insertAt(
     }
     return Either.right(false);
   }
+
+  private DirectBuffer wrapNumberOfElementInstances(
+      final MutableDirectBuffer variableBuffer,
+      final DirectBuffer variableView,
+      final long numberOfElementInstances) {
+    variableWriter.wrap(variableBuffer, 0);
+
+    variableWriter.writeInteger(numberOfElementInstances);
+    final var length = variableWriter.getOffset();
+
+    variableView.wrap(variableBuffer, 0, length);
+    return variableView;
+  }
+
+  private void updateNumberOfCompletedElementInstances(final BpmnElementContext context) {
+    updateNumberOfElementInstances(
+        context,
+        NUMBER_OF_COMPLETED_ELEMENT_INSTANCES_VARIABLE,
+        numberOfCompletedElementInstancesVariableBuffer,
+        numberOfCompletedElementInstancesVariableView);
+  }
+
+  private void updateNumberOfTerminatedElementInstances(final BpmnElementContext context) {
+    updateNumberOfElementInstances(
+        context,
+        NUMBER_OF_TERMINATED_ELEMENT_INSTANCES_VARIABLE,
+        numberOfTerminatedElementInstancesVariableBuffer,
+        numberOfTerminatedElementInstancesVariableView);
+  }
+
+  private void updateNumberOfElementInstances(
+      final BpmnElementContext context,
+      final DirectBuffer variable,
+      final MutableDirectBuffer variableBuffer,
+      final DirectBuffer variableView) {
+
+    final DirectBuffer numberOfElementInstancesBuffer =
+        stateBehavior.getLocalVariable(context, variable);
+
+    long numberOfElementInstances = 0;
+    if (numberOfElementInstancesBuffer != null) {
+      variableReader.wrap(
+          numberOfElementInstancesBuffer, 0, numberOfElementInstancesBuffer.capacity());
+      numberOfElementInstances = variableReader.readInteger();
+    }
+
+    stateBehavior.setLocalVariable(
+        context,
+        variable,
+        wrapNumberOfElementInstances(variableBuffer, variableView, numberOfElementInstances + 1));
+  }
+
+  private void updateNumberOfActiveElementInstances(final BpmnElementContext context) {
+    final ElementInstance elementInstance = stateBehavior.getElementInstance(context);
```

Comment:
🔧  We should reduce the number of read operations to the state as much as we can, because that can slow down things. We can do that by moving this call out of this method (i.e. it's already retrieved from state in `onChildActivating`, but not yet in `afterExecutionPathCompleted` and `onChildTerminated`). We can then have the value of `elementInstance.getNumberOfActiveElementInstances()` passed in as a parameter.